File: LeetCode/1530. Number of Good Leaf Nodes Pairs/1530.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int CountPairs(TreeNode root, int distance) {
        int count = 0;
        Dfs(root, distance, ref count);
        return count;
        
    }

    private List<int> Dfs(TreeNode node, int distance, ref int count) {
        if (node == null) {
            return new List<int>();
        }

        if (node.left == null && node.right == null) {
            return new List<int> { 1 };
        }

        var left = Dfs(node.left, distance, ref count);
        var right = Dfs(node.right, distance, ref count);

        foreach (var l in left) {
            foreach (var r in right) {
                if (l + r <= distance) {
                    count++;
                }
            }
        }

        var result = new List<int>();
        foreach (var l in left) {
            if (l + 1 < distance) {
                result.Add(l + 1);
            }
        }

        foreach (var r in right) {
            if (r + 1 < distance) {
                result.Add(r + 1);
            }
        }

        return result;
    }
}

/*
public class Solution {
    public int CountPairs(TreeNode root, int distance) {
        int result = 0;

        Count(root);

        return result;        

        int[] Count(TreeNode node)
        {
            if (node.left == null && node.right == null)
            {
                return new int[] { 1 };
            }

            var left = node.left == null ? null : Count(node.left);
            var right = node.right == null ? null : Count(node.right);

            int[] arr;
            if (left == null || right == null)
            {
                var leaves = left ?? right;
                arr = new int[1 + Math.Min(distance, leaves.Length)];
                for (var i = 0; i < Math.Min(arr.Length - 1, leaves.Length); ++i)
                {
                    arr[i+1] = leaves[i];
                }
            }
            else
            {
                for (var i = 0; i < left.Length; ++i)
                {
                    for (var j = 0; j < right.Length; ++j)
                    {
                        if (i + j + 2 > distance) break;

                        result += (left[i] * right[j]);                        
                    }
                }

                arr = new int[1 + Math.Min(distance, Math.Max(left.Length, right.Length))];
                for (var i = 0; i < Math.Min(left.Length, arr.Length - 1); ++i)
                {
                    arr[i + 1] += left[i];
                }

                for (var i = 0; i < Math.Min(right.Length, arr.Length - 1); ++i)
                {
                    arr[i + 1] += right[i];
                }
            }

            return arr;
        }
    }
}
*/

File: LeetCode/1530. Number of Good Leaf Nodes Pairs/1530.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countPairs(TreeNode* root, int distance) {
        int ans = 0;
        dfs(root, distance, ans);
        return ans;
        
    }

    vector<int> dfs(TreeNode* root, int distance, int& ans) {
        if (!root) return {};
        if (!root->left && !root->right) return {1};
        vector<int> left = dfs(root->left, distance, ans);
        vector<int> right = dfs(root->right, distance, ans);
        for (int l : left) {
            for (int r : right) {
                if (l + r <= distance) {
                    ans++;
                }
            }
        }
        vector<int> res;
        for (int l : left) {
            if (l + 1 < distance) {
                res.push_back(l + 1);
            }
        }
        for (int r : right) {
            if (r + 1 < distance) {
                res.push_back(r + 1);
            }
        }
        return res;
    }
};


/*
struct NodePos
{
    int64_t depth;
    int64_t pos;
};

void dfs(TreeNode *node, std::vector<NodePos> &leaves, int64_t depth, int64_t pos)
{
    if(node->left != nullptr)
    {
        dfs(node->left, leaves, depth + 1, pos);
    }

    if(node->right != nullptr)
    {
        dfs(node->right, leaves, depth + 1, pos | (int64_t(1) << depth));
    }

    if(node->left == nullptr && node->right == nullptr)
    {
        // std::cout << ((depth)) << " " << pos;
        // std::cout << std::endl;

        NodePos p;
        p.depth = depth;
        p.pos = pos;
        leaves.emplace_back(p);

        // std::cout << depth << std::endl;
    }
}

class Solution {
public:
    int countPairs(TreeNode* root, int distance)
    {
        std::vector<NodePos> leaves;

        dfs(root, leaves, 0, 0);

        int result = 0;

        std::cout << "=========" << std::endl;

        for(size_t i = 0; i < leaves.size() - 1; i++)
        {
            for(size_t j = i + 1; j < leaves.size(); j++)
            {
                auto fd = leaves[i].depth;
                auto sd = leaves[j].depth;

                auto fp = leaves[i].pos;
                auto sp = leaves[j].pos;

                auto min = std::min(fd, sd);

                // std::cout << (int)fd << " " << (int)sd << std::endl;
                // std::cout << (int64_t)fp << " " << (int64_t)sp << std::endl;

                auto depthDiff = __builtin_ctzl(fp ^ sp);
                
                auto diff = (fd - depthDiff) + sd - depthDiff;

                if(diff == 0 && fd != sd)
                {
                    diff = std::abs(fd - sd);
                }

                result += diff <= distance;
            }
        }
        return result;

    }
};
*/